File: .skills/openclaw-skills/skills/elmaslouhymouaad/searxng-search-skill/searxng_skill/retry.py
"""
Retry logic with exponential backoff
"""

import time
import functools
from typing import Callable, Any, Optional, Type, Tuple
import requests

from .exceptions import TimeoutException, ConnectionException


def retry_with_backoff(
    max_retries: int = 3,
    initial_delay: float = 1.0,
    backoff_factor: float = 2.0,
    exceptions: Tuple[Type[Exception], ...] = (
        requests.exceptions.Timeout,
        requests.exceptions.ConnectionError,
    ),
    verbose: bool = True
):
    """
    Decorator for retrying function calls with exponential backoff
    
    Args:
        max_retries: Maximum number of retry attempts
        initial_delay: Initial delay between retries in seconds
        backoff_factor: Exponential backoff multiplier
        exceptions: Tuple of exceptions to catch and retry
        verbose: Whether to print retry information
    """
    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        def wrapper(*args, **kwargs) -> Any:
            last_exception = None
            
            for attempt in range(max_retries):
                try:
                    return func(*args, **kwargs)
                    
                except exceptions as e:
                    last_exception = e
                    
                    if attempt < max_retries - 1:
                        delay = initial_delay * (backoff_factor ** attempt)
                        if verbose:
                            print(
                                f"Attempt {attempt + 1}/{max_retries} failed: {e}. "
                                f"Retrying in {delay:.2f}s..."
                            )
                        time.sleep(delay)
                    else:
                        if verbose:
                            print(f"All {max_retries} attempts failed.")
            
            raise last_exception
        
        return wrapper
    return decorator


class RetryStrategy:
    """Configurable retry strategy"""
    
    def __init__(
        self,
        max_retries: int = 3,
        initial_delay: float = 1.0,
        backoff_factor: float = 2.0,
        max_delay: float = 60.0,
        jitter: bool = True
    ):
        self.max_retries = max_retries
        self.initial_delay = initial_delay
        self.backoff_factor = backoff_factor
        self.max_delay = max_delay
        self.jitter = jitter
    
    def get_delay(self, attempt: int) -> float:
        """Calculate delay for given attempt number"""
        import random
        
        delay = self.initial_delay * (self.backoff_factor ** attempt)
        delay = min(delay, self.max_delay)
        
        if self.jitter:
            delay *= (0.5 + random.random())
        
        return delay
    
    def execute(
        self,
        func: Callable,
        *args,
        **kwargs
    ) -> Any:
        """Execute function with retry strategy"""
        last_exception = None
        
        for attempt in range(self.max_retries):
            try:
                return func(*args, **kwargs)
            except (requests.exceptions.Timeout, 
                    requests.exceptions.ConnectionError) as e:
                last_exception = e
                
                if attempt < self.max_retries - 1:
                    delay = self.get_delay(attempt)
                    time.sleep(delay)
        
        raise last_exception

File: .skills/openclaw-skills/skills/bgrober/alista/scripts/lib/utils/retry.ts
/**
 * Retry utility with exponential backoff for external API calls.
 */

import { ApiError } from "../types";

/**
 * Configuration for retry behavior
 */
export interface RetryConfig {
	/** Maximum number of retry attempts */
	maxRetries: number;
	/** Initial delay in milliseconds */
	baseDelayMs: number;
	/** Maximum delay cap in milliseconds */
	maxDelayMs: number;
	/** Function to determine if an error is retryable */
	shouldRetry: (error: unknown) => boolean;
}

/**
 * Default retry configuration
 */
export const DEFAULT_RETRY_CONFIG: RetryConfig = {
	maxRetries: 3,
	baseDelayMs: 500,
	maxDelayMs: 8000,
	shouldRetry: (error: unknown): boolean => {
		// ApiError with retryable flag
		if (error instanceof ApiError) {
			return error.retryable || (error.status >= 500 && error.status < 600);
		}

		// Standard Error - check message for common transient issues
		if (error instanceof Error) {
			const message = error.message.toLowerCase();
			return (
				error.name === "AbortError" ||
				message.includes("timeout") ||
				message.includes("econnreset") ||
				message.includes("econnrefused") ||
				message.includes("network") ||
				message.includes("fetch failed")
			);
		}

		return false;
	},
};

/**
 * Execute a function with retry logic and exponential backoff.
 *
 * @param fn - The async function to execute
 * @param config - Retry configuration (uses defaults if not provided)
 * @returns The result of the function
 * @throws The last error if all retries fail
 *
 * @example
 * ```typescript
 * const result = await withRetry(
 *   () => fetch(url, { signal: AbortSignal.timeout(10000) }),
 *   { maxRetries: 3, baseDelayMs: 500 }
 * );
 * ```
 */
export async function withRetry<T>(
	fn: () => Promise<T>,
	config: Partial<RetryConfig> = {},
): Promise<T> {
	const finalConfig: RetryConfig = { ...DEFAULT_RETRY_CONFIG, ...config };
	let lastError: unknown;

	for (let attempt = 0; attempt <= finalConfig.maxRetries; attempt++) {
		try {
			return await fn();
		} catch (error) {
			lastError = error;

			// Check if we should retry
			if (!finalConfig.shouldRetry(error) || attempt === finalConfig.maxRetries) {
				throw error;
			}

			// Calculate delay with exponential backoff and jitter
			const baseDelay = finalConfig.baseDelayMs * 2 ** attempt;
			const jitter = Math.random() * 0.2 * baseDelay; // 0-20% jitter
			const delay = Math.min(baseDelay + jitter, finalConfig.maxDelayMs);

			console.log(
				`Retry attempt ${attempt + 1}/${finalConfig.maxRetries} after ${delay.toFixed(0)}ms`,
			);

			await sleep(delay);
		}
	}

	throw lastError;
}

/**
 * Sleep for a given number of milliseconds
 */
function sleep(ms: number): Promise<void> {
	return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Create a retryable fetch wrapper with timeout
 *
 * @param url - URL to fetch
 * @param options - Fetch options
 * @param timeoutMs - Timeout in milliseconds
 * @param retryConfig - Optional retry configuration
 * @returns Response
 *
 * @example
 * ```typescript
 * const response = await retryableFetch(
 *   'https://api.example.com/data',
 *   { method: 'GET' },
 *   10000
 * );
 * ```
 */
export async function retryableFetch(
	url: string,
	options: RequestInit,
	timeoutMs: number,
	retryConfig: Partial<RetryConfig> = {},
): Promise<Response> {
	return withRetry(async () => {
		const response = await fetch(url, {
			...options,
			signal: AbortSignal.timeout(timeoutMs),
		});

		// Convert HTTP errors to ApiError for proper retry handling
		if (!response.ok) {
			const isRetryable = response.status >= 500 && response.status < 600;
			throw new ApiError(
				`HTTP ${response.status}: ${response.statusText}`,
				response.status,
				isRetryable,
			);
		}

		return response;
	}, retryConfig);
}
